tests/test_core_utils.py

```diff
@@ -28,15 +28,14 @@ def test_accessing_hydra_config(hydra_restore_singletons: Any) -> Any:
     assert cfg.config_name == "accessing_hydra_config"
 
 
-def test_py_version_resolver(hydra_restore_singletons: Any, monkeypatch: Any) -> Any:
+def test_py_version_resolver(
+    hydra_restore_singletons: Any, monkeypatch: Any, restore_resolvers: Any
+) -> Any:
     Version = namedtuple("Version", "major minor micro")
-    with monkeypatch.context() as m:
-        m.setattr(sys, "version_info", Version(3, 7, 2))
-        OmegaConf.clear_resolvers()
-        utils.setup_globals()
-        assert OmegaConf.create({"key": "${python_version:}"}).key == "3.7"
-        assert OmegaConf.create({"key": "${python_version:major}"}).key == "3"
-        assert OmegaConf.create({"key": "${python_version:minor}"}).key == "3.7"
-        assert OmegaConf.create({"key": "${python_version:micro}"}).key == "3.7.2"
-    # reset OmegaConf resolver with no patching
+    monkeypatch.setattr(sys, "version_info", Version(3, 7, 2))
+    OmegaConf.clear_resolvers()
     utils.setup_globals()
```

Comment:
I don't think you need those two lines now.

Reply:
Still need them :)

```python
$ git diff
diff --git a/tests/test_core_utils.py b/tests/test_core_utils.py
index 5dbf3e85..ae94e01f 100644
--- a/tests/test_core_utils.py
+++ b/tests/test_core_utils.py
@@ -33,8 +33,8 @@ def test_py_version_resolver(
 ) -> Any:
     Version = namedtuple("Version", "major minor micro")
     monkeypatch.setattr(sys, "version_info", Version(3, 7, 2))
-    OmegaConf.clear_resolvers()
-    utils.setup_globals()
+    # OmegaConf.clear_resolvers()
+    # utils.setup_globals()
     assert OmegaConf.create({"key": "${python_version:}"}).key == "3.7"
     assert OmegaConf.create({"key": "${python_version:major}"}).key == "3"


$ pytest tests/test_core_utils.py 
=============================================================================== test session starts ================================================================================
platform darwin -- Python 3.8.2, pytest-5.4.2, py-1.8.1, pluggy-0.13.1
rootdir: /Users/jieru/workspace/hydra-fork/hydra, inifile: pytest.ini
plugins: snail-0.1.0, hydra-core-1.0.0rc3
collected 2 items                                                                                                                                                                  

tests/test_core_utils.py .F                                                                                                                                                  [100%]

===================================================================================== FAILURES =====================================================================================
_____________________________________________________________________________ test_py_version_resolver _____________________________________________________________________________

hydra_restore_singletons = None, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fea682d7b50>, restore_resolvers = None

    def test_py_version_resolver(
        hydra_restore_singletons: Any, monkeypatch: Any, restore_resolvers: Any
    ) -> Any:
        Version = namedtuple("Version", "major minor micro")
        monkeypatch.setattr(sys, "version_info", Version(3, 7, 2))
        # OmegaConf.clear_resolvers()
        # utils.setup_globals()
>       assert OmegaConf.create({"key": "${python_version:}"}).key == "3.7"
E       AssertionError: assert '3.8' == '3.7'
E         - 3.7
E         + 3.8

tests/test_core_utils.py:38: AssertionError
============================================================================= short test summary info ==============================================================================
FAILED tests/test_core_utils.py::test_py_version_resolver - AssertionError: assert '3.8' == '3.7'
=========================================================================== 1 failed, 1 passed in 0.25s ============================================================================
(pytest38) jieru-mbp:hydra jieru$ 

```

Reply:
Why? could this be related to the OmegaConf cache?
try OmegaConf.clear_cache()

Reply:
sure i can try that. [`clear_cache`](https://github.com/omry/omegaconf/blob/a6fde5a748be264f7d8344bc086e0b7aa237cb5d/omegaconf/omegaconf.py#L378) expects a  `conf` param, what should that be in this case? 

I think the reason that it fails is there is an another test `test_accessing_hydra_config` that runs first before this one, and as a result it already register the python_version function before this test run. Since OmegaConf won't let `setup_globals` override the resolver, the monkeypatched function never gets called. as a result, i think these 2 lines of code are needed.

Reply:
That's actually an error in the API. the cache it's clearing is static so it should not require a config.

Generally speaking, tests should clear after themselves and not before they run.
not sure how hard it is to fix in the current state of things.

Reply:
Yea you are right, this does feel weird. Instead of cleaning up before the run in this test, I've edited the other tests to clean up after themselves.

Reply:
Did some further digging. I think we need `clear_resolver` instead of `restore`. Added a fixture for that instead.

Reply:
The idea with tests is that they are not supposed to change the state of the system.
whatever was before, should stay.  Your fix is sweeping the problem under the rug.
It can be a bit tricky to find which test is changing the state of the system. one way I have been using is to  run multiple tests at one, with the test that triggers the problem last. by running different test files before the one the blows up you can find which one is causing issues.